Configuring Session Certificates for Dynamic Networks (#754)
corda/corda-docs-portal: content/en/platform/corda/5.0-alpha/deploying/deployment-tutorials/session-certificates.md

---
date: '2022-11-15'
title: "Configuring Optional Session Certificates"
menu:
  corda-5-alpha:

>     identifier: corda-5-alpha-mgm-onboarding

Reviewer: I'm not too familiar with these headers but these seems like it's referencing the wrong page

Author: > I'm not too familiar with these headers but these seems like it's referencing the wrong page

good spot, thanks!

## Set Variables

* Set the holding identity short hash of the virtual node of either the MGM or member as a variable for use in later commands:
```shell
export HOLDING_ID=<holding-id>
```
<!--add links to two section where holding ID is retrieved?-->

## Generate a Certificate Signing Request (CSR)

After creating the MGM or member session key pair, but before building the registration context, generate a CSR for the session certificate by running the following command, replacing `X500_NAME` with the X500Name of the MGM or member:
```shell
curl --fail-with-body -s -S -k -u admin:admin  -X POST -H "Content-Type: application/json" -d '{"x500Name": "'$X500_NAME'"}' $API_URL"/certificates/"$HOLDING_ID/$SESSION_KEY_ID > $WORK_DIR/request.csr
```
Similarly to the TLS certificate, the CSR can be processed to issue a certificate using either a real CA of your choice or the fake CA dev tool. To use the fake CA tool:
```shell
cd $RUNTIME_OS
java -jar ./applications/tools/p2p-test/fake-ca/build/bin/corda-fake-ca-5.0.0.0-SNAPSHOT.jar -m /tmp/ca csr $WORK_DIR/request.csr
cd $WORK_DIR
```
This command outputs the location of the signed certificate. For example:
```shell

> Wrote certificate to /tmp/ca/request1/certificate.pem

Reviewer: I think we should remove mention of the fake CA from the public docs.

```
Similarly to the TLS certificate, the CSR can be processed to issue a certificate using a CA chosen by the MGM operator. The CA trustroot for session certificates should be configured during the MGM onboarding.
```

```

> At this point, you should already have a certificate based on the CSR exported from Corda issued either by a real CA or by the fake CA tool. You must upload the certificate chain to the Corda cluster. To upload the certificate chain, run:

Reviewer: ```suggestion
Once you have a certificate based on the CSR exported from Corda issued by the CA, you must upload the certificate chain to the Corda cluster. To upload the certificate chain, run:
```

```shell
curl -k -u admin:admin -X PUT  -F certificate=@/tmp/ca/request/certificate.pem -F alias=session-certificate $API_URL/certificates/vnode/$HOLDING_ID/p2p-session
```
You can optionally omit the root certificate.

{{< note >}}
If you upload a certificate chain consisting of more than one certificates, ensure that `-----END CERTIFICATE-----` and `-----BEGIN CERTIFICATE-----` from the next certificate are separated by a new line with no empty spaces in between.
{{< /note >}}

### Disable Revocation Checks

If the CA has not been configured with revocation (for example, via CRL or OCSP), you can disable revocation checks. By default, revocation checks are enabled.

> The fake CA dev tool does not support revocation and so, if you are using the fake CA, you must disable revocation checks.

Reviewer: ```suggestion

```

To disable revocation checks, do the following:
1. Retrieve the current link manager configuration version:
   ```shell
   curl --insecure -u admin:admin -X GET $API_URL/config/corda.p2p.linkManager
   ```
2. Save the displayed version number from the response as a variable:
   ```shell
   export CONFIG_VERSION=<configuration-version>
   ```
3. Send the following request to disable revocation checks for the specified gateway worker:
   ```
   curl -k -u admin:admin -X PUT -d '{"section":"corda.p2p.linkManager", "version":"'$CONFIG_VERSION'", "config": "{ "revocationCheck": { "mode": "OFF" } }", "schemaVersion": {"major": 1, "minor": 0}}' $API_URL"/config"
   ```

> ## Build Registration Context for MGM Registration

Reviewer: I would move this section to the top after `## Set Variables`.
This section is in regards to the MGM onboarding, whereas the rest of the doc is in regards to member onboarding. 
It think we should link to this from the "build registration context" of the MGM onboarding doc also.

If using session certificates, you must also specify the following in the [MGM REGISTRATION_CONTEXT](mgm-onboarding.html#build-registration-context):

> * Add an extra JSON field `corda.group.truststore.session.0` with the truststore of the CA to the registration context (similar to `corda.group.truststore.tls.0`).

Reviewer 2: Should these two bullets points be numbered bullets?

* Set the JSON field `corda.group.pki.session` to `"Standard"` instead of `"NoPKI"`.

## Configure Virtual Node as Network Participant

> If using session certificates, you must also add the `sessionCertificateChainAlias` JSON field to the network setup RPC request. For example:

Reviewer: ```suggestion
If using session certificates, you must also add the `sessionCertificateChainAlias` and `useClusterLevelSessionCertificateAndKey` JSON fields to the network setup RPC request. For example:
```
